server/src/test/java/org/elasticsearch/index/query/FirstQueryBuilderTests.java

```diff
@@ -0,0 +1,89 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.query;
+
+import org.apache.lucene.search.Query;
+import org.elasticsearch.index.query.first.FirstQuery;
+import org.elasticsearch.test.AbstractQueryTestCase;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.hamcrest.CoreMatchers.instanceOf;
+
+public class FirstQueryBuilderTests extends AbstractQueryTestCase<FirstQueryBuilder> {
+
+    @Override
+    protected FirstQueryBuilder doCreateTestQueryBuilder() {
+        int numClauses = randomIntBetween(1, 5);
+        List<QueryBuilder> queryBuilders = new ArrayList<>(numClauses);
+        for (int i = 0; i < numClauses; i++) {
+            queryBuilders.add(RandomQueryBuilder.createQuery(random()));
+        }
+        FirstQueryBuilder queryBuilder = new FirstQueryBuilder(queryBuilders);
+        return queryBuilder;
+    }
+
+    @Override
+    protected void doAssertLuceneQuery(FirstQueryBuilder queryBuilder, Query query, QueryShardContext context) throws IOException {
+        assertThat(query, instanceOf(FirstQuery.class));
+        FirstQuery firstQuery = (FirstQuery) query;
+        assertEquals(queryBuilder.getQueryBuilders().size(), firstQuery.getQueries().length);
+    }
+
+    public void testFromJson() throws IOException {
+        String query =
+            "{" +
+                "\"first\" : {" +
+                "  \"queries\" : [ {" +
+                "    \"term\" : {" +
+                "      \"tag\" : {" +
+                "        \"value\" : \"white\"," +
+                "        \"boost\" : 1.0" +
+                "      }" +
+                "    }" +
+                "  }, {" +
+                "    \"term\" : {" +
+                "      \"tag\" : {" +
+                "        \"value\" : \"elephant\"," +
+                "        \"boost\" : 1.0" +
+                "      }" +
+                "    }" +
+                "  } ]," +
+                "  \"boost\" : 42.0" +
+                "}" +
+            "}";
+
+        FirstQueryBuilder queryBuilder = (FirstQueryBuilder) parseQuery(query);
+        checkGeneratedJson(query, queryBuilder);
+        assertEquals(query, 42, queryBuilder.boost, 0.00001);
+    }
+
+    public void testRewrite() throws IOException {
+        QueryBuilder queryBuilder = new WrapperQueryBuilder(new TermsQueryBuilder("field", "elephant").toString());
+        List<QueryBuilder> queryBuilders = new ArrayList<>(Arrays.asList(queryBuilder));
+        FirstQueryBuilder firstQueryBuilder = new FirstQueryBuilder(queryBuilders);
+        FirstQueryBuilder rewritten = (FirstQueryBuilder) firstQueryBuilder.rewrite(createShardContext());
+        assertNotSame(rewritten, firstQueryBuilder);
+        assertEquals(new TermsQueryBuilder("field", "elephant"), rewritten.getQueryBuilders().get(0));
```

Comment:
Also after merging in master this fails, I with a cast exception, I think because meanwhile we rewrite queries with unmapped fields to MatchNoDocs. Might be wrong here, but worth checking.

Reply:
Addressed in 5fba133